### PR TITLE
Add metrics cardinality control for Application Signals

### PR DESCRIPTION
**Pull request by mxiamxia**
This PR is an updated version from https://github.com/aws/amazon-cloudwatch-agent/pull/1000, Since @bjrara is not available atm so I create this new PR with a few changes below. 
1. Removed Metrics labels referenece in `MetricData` structure
2. Disabled Metric Limiter by default until AppSignals console team completed the related changes for supporting E2E CX.
# Description of the issue
Due to the nature of metrics generated from auto-instrumentation, high cardinality metrics may occur if attributes which are intended to have low cardinality, are assigned values with high cardinality.

# Description of changes
A new component `metric_limiter` is introduced into `awsappsignalsprocessor` to enhance cardinality control. Customers can set `drop_threshold` to indicate the anticipated number of low cardinality metrics for each `service`. The local metrics limiter will prioritize exporting the most crucial metrics (e.g. most frequently used, manually retained) up to that threshold. Metrics identified as high cardinality will be rolled up, and only the number of observations will be reported.

# License
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

# Tests
1. Unit tests.
3. E2E tests.
4. Performance tests.

# Requirements
_Before commit the code, please do the following steps._
1. Run `make fmt` and `make fmt-sh`
5. Run `make lint`

### CODE REVIEW

File: plugins/processors/awsappsignals/rules/keeper.go [hunk 1]
```diff
@@ -3,15 +3,21 @@
...
+import (
+	"go.opentelemetry.io/collector/pdata/pcommon"
+
+	"github.com/aws/amazon-cloudwatch-agent/plugins/processors/awsappsignals/common"
+)
 
 type KeepActions struct {
-	Actions []ActionItem
+	Actions                 []ActionItem
+	markDataPointAsReserved bool
```
Comment (lisguo) [inline plugins/processors/awsappsignals/rules/keeper.go:14]: Can you remind me what `markDataPointAsReserved` does? 

I see that we pass in `!limiterConfig.Disabled` to this value... not sure I completely understand

File: plugins/processors/awsappsignals/internal/cardinalitycontrol/count_min_sketch.go [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+	for i := 0; i < cms.depth; i++ {
+		hashFunc := cms.hashFuncs[i]
+		hashValue := hashFunc(obj.HashKey())
+		pos := int(hashValue % int64(cms.width))
+
+		cms.matrix[i][pos] += obj.Frequency()
+	}
+}
+
+func NewCountMinSketch(depth, width int, hashFuncs ...CountMinSketchHashFunc) *CountMinSketch {
```
[the remark this answers is not in the document]
Comment (mxiamxia) [reply plugins/processors/awsappsignals/internal/cardinalitycontrol/count_min_sketch.go:37]: It is being used under `MetircLimiter.go::newService()` function. so we need to create a new CMS probability structure for a new individual micro-service. :)

File: plugins/processors/awsappsignals/rules/keeper.go (hunk 1 above)
Comment (mxiamxia) [reply plugins/processors/awsappsignals/rules/keeper.go:14]: Yeah, it is to make sure the metrics that `Replaced` or `Kept` by the custom rules will not be aggregated by MetricLimiter which happens later.  

So if MetricLimiter is enabled, we will inject an placeholder `markDataPointAsReserved` attribute to Metrics, and the MetricLimiter will not process such metrics later. If MetricLimiter is not enabled in CWAgent, we just don't do anything.

File: plugins/processors/awsappsignals/config/config.go [hunk 3]
```diff
@@ -4,14 +4,47 @@
...
+	Limiter   *LimiterConfig `mapstructure:"limiter"`
+}
+
+type LimiterConfig struct {
+	Threshold                 int             `mapstructure:"drop_threshold"`
+	Disabled                  bool            `mapstructure:"disabled"`
+	LogDroppedMetrics         bool            `mapstructure:"log_dropped_metrics"`
+	RotationInterval          time.Duration   `mapstructure:"rotation_interval"`
+	GarbageCollectionInterval time.Duration   `mapstructure:"-"`
+	ParentContext             context.Context `mapstructure:"-"`
```
Comment (lisguo) [inline plugins/processors/awsappsignals/config/config.go:26]: Can you describe a use case where we will need to use the parent context?

File: plugins/processors/awsappsignals/internal/cardinalitycontrol/metrics_limiter.go [hunk 4]
```diff
@@ -0,0 +1,415 @@
...
+			svc = newService(serviceName, m.DropThreshold, m.RotationInterval, m.ctx, m.logger)
+			m.services[serviceName] = svc
+		}
+		m.mapLock.Unlock()
+	}
+
+	metricData := newMetricData(serviceName, metricName, labels)
+
+	reserved, _ := attributes.Get(common.AttributeTmpReserved)
+	if reserved.Bool() {
```
[the remark this answers is not in the document]
Comment (mxiamxia) [reply plugins/processors/awsappsignals/internal/cardinalitycontrol/metrics_limiter.go:115]: Yes, in the case of MetricLimiter is enabled, but no metric being matched from Keeper or Replacer rules (or no rule being defined). These metrics will have `reserved.Bool()` is false.

File: plugins/processors/awsappsignals/config/config.go (hunk 3 above)
Comment (mxiamxia) [reply plugins/processors/awsappsignals/config/config.go:26]: Yes, we inherit the `ParentContext` to make sure the following new GoRoutine can be properly returned. 

```
func NewMetricsLimiter(config *config.LimiterConfig, logger *zap.Logger) Limiter {
	logger.Info("creating metrics limiter with config", zap.Any("config", config))

	ctx := config.ParentContext
	if ctx == nil {
		ctx = context.TODO()
	}

	limiter := &MetricsLimiter{
		DropThreshold:     config.Threshold,
		LogDroppedMetrics: config.LogDroppedMetrics,
		RotationInterval:  config.RotationInterval,

		logger:   logger,
		ctx:      ctx,
		services: map[string]*service{},
	}

	go func() {
		for {
			select {
			case <-ctx.Done():
				return
			default:
				limiter.removeStaleServices()
				time.Sleep(config.GarbageCollectionInterval)
			}
		}
	}()
```

**Comment by mxiamxia** [comment]
Rebased!

**Review by pxaws — APPROVED** [review]

File: plugins/processors/awsappsignals/common/types.go [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+	MetricAttributeLocalOperation  = "Operation"
+	MetricAttributeRemoteService   = "RemoteService"
+	MetricAttributeRemoteOperation = "RemoteOperation"
+	MetricAttributeRemoteTarget    = "RemoteTarget"
+)
+const (
+	HostedInAttributeClusterName  = "HostedIn.EKS.Cluster"
+	HostedInAttributeK8SNamespace = "HostedIn.K8s.Namespace"
+	HostedInAttributeEnvironment  = "HostedIn.Environment"
+)
```
Comment (majanjua-amzn) [inline plugins/processors/awsappsignals/common/types.go:23]: Do we not also want `HostedIn.K8s.Cluster` now?

**Review by majanjua-amzn — APPROVED** [review]

File: plugins/processors/awsappsignals/common/types.go (hunk 5 above)
Comment (majanjua-amzn) [reply plugins/processors/awsappsignals/common/types.go:23]: Followed up offline, team will take a followup action to add any attributes needed for new platform support

**Pull request merged**